### PR TITLE
18: Post Sitemaps

### LINKED ISSUE #18: Posts Sitemap
The Sitemap must:

Begin with an opening <urlset> tag and end with a closing </urlset> tag.
Specify the namespace (protocol standard) within the <urlset> tag.
Include a <url> entry for each URL, as a parent XML tag.
Include a <loc> child entry for each <url> parent tag.

https://www.sitemaps.org/protocol.html

Implementation should be informed by performance considerations made previously in #1

Should paginate itself sequentially.

Not included:
Caching
Filters
Post status filters

### PR DESCRIPTION
**Pull request by svandragt**
### Issue Number
Fixes #18
Supersedes #28  and #31  

### Description
Post sitemap provider, with pagination. 2000 posts per page. The distribution strategy discussed in #39 is outwith the scope of this PR.

### Screenshots (before and after if applicable)

image

### Type of change
Please select the relevant options:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Steps to test
- Displays page 1: http://sitemaps.local/sitemap.xml?sitemap=posts 
- Page 2: http://sitemaps.local/sitemap.xml?sitemap=posts&paged=2

### Acceptance criteris
- [x] My code follows WordPress coding standards.
- [x] I have performed a self-review of my own code.
- [x] If the changes are visual, I have cross browser / device tested.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [] I have added test instructions that prove my fix is effective or that my feature works.

### CODE REVIEW

**Comment by svandragt** [comment]
@googlebot I consent.

File: inc/class-sitemaps-index.php
Comment (joemcgill) [inline inc/class-sitemaps-index.php]: Let's move all hook callbacks for this sitemap into a `bootstrap()` method inside the `Core_Sitemaps_Post` class and call that method once here instead of adding hooks and filters in the sitemaps index.

File: inc/class-sitemaps-posts.php
Comment (joemcgill) [inline inc/class-sitemaps-posts.php]: Let's set our maximum posts_per_page as a plugin level constant for now rather than calculating them during the query.

File: inc/class-sitemaps-posts.php
Comment (joemcgill) [inline inc/class-sitemaps-posts.php]: For sitemap pages, the main XML node should be:
`<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">`

See: https://www.sitemaps.org/protocol.html

File: inc/class-sitemaps-posts.php
Comment (joemcgill) [inline inc/class-sitemaps-posts.php]: We're not really returning anything here since we're printing output directly and then exiting, so no return value is needed.

File: inc/class-sitemaps-index.php
Comment (joemcgill) [inline inc/class-sitemaps-index.php]: I'd like to instantiate one instance of the registry in the plugin and reference that instance rather than relying on the singleton pattern here. For now, let's make this a public property of this class and instantiate the registry in the constructor, but we might end up moving it to a base `Core_Sitemaps` class in the future.

File: inc/class-sitemaps-posts.php
Comment (joemcgill) [inline inc/class-sitemaps-posts.php]: Let's fire this on the `core_sitemaps_setup_sitemaps` action so we're registering sitemaps just before the registry adds the rewrites. We should also try name our functions/methods so they describe the action being performed. In this case, something like `register_sitemap()` or `setup_sitemap()` would be more clear.

File: inc/class-sitemaps-posts.php
Comment (joemcgill) [inline inc/class-sitemaps-posts.php]: This method name could also be more clear if it described the action it performs, i.e., `render_sitemap()`.

File: inc/class-sitemaps-posts.php
Comment (joemcgill) [inline inc/class-sitemaps-posts.php]: Let's not save this content to a property just yet since we're not referencing it anywhere else and will eventually want this class to support multiple post types, which would cause clashes if we're overwriting the content property for each post type and page.

File: inc/class-sitemaps-posts.php
Comment (joemcgill) [inline inc/class-sitemaps-posts.php]: Eventually, we'll need to support multiple post types with this method. Let's have it accept two parameters: `$post_type` and `$page_num`.

File: inc/class-sitemaps-index.php
Comment (joemcgill) [inline inc/class-sitemaps-index.php]: We can completely remove the `return` value here.

**Review by joemcgill — CHANGES_REQUESTED** [review]
The output for the first page here is working and I like the general direction. I've left a bunch of initial feedback notes. I'm not super worried about us supporting multiple post types in this first run, but we should be able to support demonstrate support for multiple pages before we merge this PR.

File: inc/class-sitemaps-index.php
Comment (svandragt) [reply inc/class-sitemaps-index.php]: We need to have discuss an approach where sitemap providers can register themselves without adding any code to other classes, this is where a Core_Sitemap class is needed, and is making things more difficult, so that would be a good next task.

Comment (svandragt) [reply inc/class-sitemaps-index.php]: If we're not using the index class as the main class, then we should probably move the bootstrap call and the registry instantiation into the main core-sitemaps.php file, later replaced by the Core_Sitemaps class there, and keep it out of Index altogether.

I also think replacing a line that explains where something is hooked into something and what it does with a call to a generic method (bootstrap) is not clearer to reason about.

Comment (svandragt) [reply inc/class-sitemaps-index.php]: What about this in core_sitemaps.php?
```php
function core_sitemaps_bootstrap() {
	$core_sitemaps_index = new Core_Sitemaps_Index();
	$core_sitemaps_index->bootstrap();

	$core_sitemaps_posts = new Core_Sitemaps_Posts();
	$core_sitemaps_posts->bootstrap();

	// Setup all registered sitemap data providers, after all others.
	$registry = Core_Sitemaps_Registry::instance();
	add_action( 'init', array( $registry, 'setup_sitemaps' ), 100 );
};

core_sitemaps_bootstrap();
```
it avoids adding anything to the global namespace and can be replaced with a class later. It is not compatible with calling the registry property of the Core_Sitemaps_Index class from the Core_Sitemaps_Posts class https://github.com/GoogleChromeLabs/wp-sitemaps/pull/33#discussion_r340357084 though.

File: inc/class-sitemaps-posts.php
Comment (svandragt) [reply inc/class-sitemaps-posts.php]: However the issue here was I wasn't using the template_redirect filter. Now that I switched to that your comment applies.

File: inc/class-sitemaps-index.php
Comment (svandragt) [reply inc/class-sitemaps-index.php]: Nevermind the previous comment, it doesn't matter for now as it will get replaced with a Core_Sitemaps class.

File: inc/class-sitemaps-index.php
Comment (svandragt) [reply inc/class-sitemaps-index.php]: No, wee need to return $template if the sitemap get variable isn't posts, otherwise the rest of the site goes blank.

**Comment by svandragt** [comment]
@joemcgill All feedback addressed. Also I think you might need to consent again on this PR.

File: inc/class-sitemaps-posts.php
Comment (svandragt) [reply inc/class-sitemaps-posts.php]: I've also adjusted the method in the index class for consistency.

File: inc/class-sitemaps-posts.php
Comment (svandragt) [reply inc/class-sitemaps-posts.php]: Done the same on the index class.

File: inc/class-sitemaps-posts.php
Comment (joemcgill) [inline inc/class-sitemaps-posts.php]: This should be `add_action` for `template_redirect` rather than `add_filter`.

File: inc/class-sitemaps-posts.php
Comment (kirstyburgoine) [inline inc/class-sitemaps-posts.php:32]: Do we also need to store a date modified value as an $arg here as well? On the sitemaps index we need to display the `<loc>` and `<lastmod>` values

Comment (joemcgill) [reply inc/class-sitemaps-posts.php:32]: Eventually, we'll need to be able to get the lastmod from each sitemap page, but let's not worry about that here. For now, just being able to see the sitemap rendered out will be fine.

**Review by joemcgill — CHANGES_REQUESTED** [review]
This is looking good Sander, I can see `sitemap-post.xml` rendering correctly now. I've got one small comment and I think this can merge.

**Comment by joemcgill** [comment]
@googlebot I consent.

File: inc/class-sitemaps-posts.php
Comment (svandragt) [reply inc/class-sitemaps-posts.php]: Added.

File: inc/class-sitemaps-posts.php
Comment (svandragt) [reply inc/class-sitemaps-posts.php:32]: There's no link here or mechanism to the page object to get the last modified date from (either way) so currently that's not possible.

**Comment by svandragt** [comment]
@joemcgill remaining feedback addressed. The distribution strategy discussed in #39 is outwith the scope of this PR and will be progressed in another PR. can merge this after approval.

**Review by joemcgill — APPROVED** [review]
Looks like this needs merge conflicts resolved, but my feedback is addressed. Pre-approving.

**Pull request merged**
